File: day1/excercise 1/java/Product.java
public class Product {
    private final String productName;
    private double stock;

    public Product(String productName) {
        this(productName, 0);
    }

    public Product(String productName, double initialStock) {
        if (productName == null || productName.isBlank()) {
            throw new IllegalArgumentException("Product name cannot be empty.");
        }

        if (initialStock < 0) {
            throw new IllegalArgumentException("Initial stock cannot be negative.");
        }

        this.productName = productName;
        this.stock = initialStock;
    }

    public String getProductName() {
        return productName;
    }

    public double getStock() {
        return stock;
    }

    public void addStock(double quantity) {
        if (quantity <= 0) {
            throw new IllegalArgumentException("Quantity must be greater than zero.");
        }

        stock += quantity;
    }

    public void sell(double quantity) {
        if (quantity <= 0) {
            throw new IllegalArgumentException("Quantity must be greater than zero.");
        }

        if (quantity > stock) {
            throw new IllegalStateException("Insufficient stock.");
        }

        stock -= quantity;
    }
}

File: day1/excercise 1/python/Product.py
class Product:
    def __init__(self, product_name: str, initial_stock: float = 0):
        if not product_name or not product_name.strip():
            raise ValueError("Product name cannot be empty.")

        if initial_stock < 0:
            raise ValueError("Initial stock cannot be negative.")

        self._product_name = product_name
        self._stock = initial_stock

    @property
    def product_name(self) -> str:
        return self._product_name

    @property
    def stock(self) -> float:
        return self._stock

    def add_stock(self, quantity: float) -> None:
        if quantity <= 0:
            raise ValueError("Quantity must be greater than zero.")

        self._stock += quantity

    def sell(self, quantity: float) -> None:
        if quantity <= 0:
            raise ValueError("Quantity must be greater than zero.")

        if quantity > self._stock:
            raise ValueError("Insufficient stock.")

        self._stock -= quantity